resolwe/flow/managers/listener/metrics.py

```diff
@@ -0,0 +1,179 @@
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+from opentelemetry.sdk.resources import Resource
+
+from resolwe.flow.executors.socket_utils import (
+    Message,
+    MessageProcessingCallback,
+    MessageProcessingEventType,
+)
+
+logger = logging.getLogger(__name__)
+
+# How often to export metrics (in milliseconds).
+METRICS_EXPORT_INTERVAL = 60000
+# TODO: Remove this when metrics are properly implemented.
+METRICS_INSECURE_EXPORT = True
+
+
+class MetricsEventReporter(MessageProcessingCallback):
+    """Class handling the reporting of the metrics events."""
+
+    # The prefix used in all metric names.
+    metrics_prefix = "resolwe_listener_"
+
+    def __init__(self):
+        """Initialize the metrics event reporter."""
+        if (endpoint := os.environ.get("METRICS_ENDPOINT")) is not None:
+            self._enabled = False
+            logger.warning(
+                "Listener metrics endpoint is not set, diabling metrics reporting."
+            )
+        else:
+            self._enabled = True
+            self._init_metrics(endpoint)
+
+    def _init_metrics(self, metric_endpoint):
+        # Initialize the counters for the metrics.
+        self._processing_messages = 0
+        self._queued_messages = 0
+
+        # Create the dictionaries with the appropriate timestamps.
+        self._processing_started: dict[bytes, float] = {}
+        self._message_received: dict[bytes, float] = {}
+
+        # Initialize the opentelemetry metrics.
+        metric_exporter = OTLPMetricExporter(
+            endpoint=metric_endpoint, insecure=METRICS_INSECURE_EXPORT
+        )
+        metric_reader = PeriodicExportingMetricReader(
+            metric_exporter, METRICS_EXPORT_INTERVAL
+        )
+        provider = MeterProvider(
+            metric_readers=[metric_reader],
+            resource=Resource.create({"service.name": "listener"}),
+        )
+
+        # Sets the global default meter provider
+        metrics.set_meter_provider(provider)
+
+        # Creates a meter from the global meter provider
+        meter = metrics.get_meter("listener")
+
+        # Create the metrics.
+        self._messages_queued = meter.create_up_down_counter(
+            name=f"{self.metrics_prefix}_messages_in_queue",
+            description="Number of unprocessed messages in the queue.",
+        )
+        self._messages_processing = meter.create_up_down_counter(
+            name=f"{self.metrics_prefix}_messages_processing",
+            description="Number of currently processing messages.",
+        )
+        self._preparation_time_histogram = meter.create_histogram(
+            name=f"{self.metrics_prefix}_preparation_duration",
+            description="Response time for messages.",
+            unit="seconds",
+        )
+        self._response_time_histogram = meter.create_histogram(
+            name=f"{self.metrics_prefix}_response_time_duration",
```

Comment:
I would use either `time` or `duration`, not `time_duration` because it seems redundant. Note that in grafana, `seconds` (unit name), will be appended as well.

Reply:
OK, I have decided on `duration`.